src/core/importDataSources.ts

```diff
@@ -0,0 +1,190 @@
+import { URL } from 'whatwg-url';
+import { getFileMimeType } from '../io';
+import { readRemoteManifestFile } from '../io/manifest';
+import { ARCHIVE_FILE_TYPES } from '../io/mimeTypes';
+import { extractFilesFromZip } from '../io/zip';
+import Pipeline, { Handler } from './pipeline';
+import { canFetchUrl, fetchFile } from '../utils/fetch';
+
+/**
+ * Represents a URI source with a file name for the downloaded resource.
+ */
+export interface UriSource {
+  uri: string;
+  name: string;
+}
+
+/**
+ * Represents a user-specified file.
+ */
+export interface FileSource {
+  file: File;
+  fileType: string;
+}
+
+/**
+ * If an archive source is specified, then it is assumed that the data source
+ * has a FileSource (representing the file inside the archive), and a parent
+ * data source that has a FileSource of the containing archive.
+ */
+export interface ArchiveSource {
+  // Full path + filename inside the archive
+  path: string;
+}
+
+/**
+ * Cases to handle:
+ *   associated dataset file is part of a (potentially nested) remote archive
+ */
+export interface DataSource {
+  uriSrc?: UriSource;
+  fileSrc?: FileSource;
+  archiveSrc?: ArchiveSource;
+  parent?: DataSource;
+}
+
+function isArchive(r: DataSource): r is DataSource & { fileSrc: FileSource } {
+  return !!r.fileSrc && ARCHIVE_FILE_TYPES.has(r.fileSrc.fileType);
+}
+
+/**
+ * Transforms a file data source to have a mime type
+ * @param dataSource
+ */
+const retypeFile: Handler<DataSource> = async (dataSource, { next }) => {
+  let src = dataSource;
+  const { fileSrc } = src;
+  if (fileSrc && fileSrc.fileType === '') {
+    const mime = await getFileMimeType(fileSrc.file);
+    if (mime) {
+      src = {
+        ...src,
+        fileSrc: {
+          ...fileSrc,
+          fileType: mime,
```

Comment:
Do we want to assign the `File.type` with the found mime string and remove this property?

Reply:
Probably, if we know we won't care to change `file.type` more than once. Assigning `fileType` is much easier than `file.type`, which is readonly and requires doing `file = new File([file], file.name, { type: newType })`. Having `fileType` just makes things easier, even if it's a redundant property slot.

Reply:
I actually have a a possible reason to not re-assign the `file` property: it might be the case that we care about matching pipeline return values with File inputs by comparing the File instance. I'm opting to keep the `fileType` property for now.